vm/src/stdlib/io.rs

```diff
@@ -442,19 +442,65 @@ fn text_io_wrapper_init(vm: &VirtualMachine, args: PyFuncArgs) -> PyResult {
 fn text_io_base_read(vm: &VirtualMachine, args: PyFuncArgs) -> PyResult {
     arg_check!(vm, args, required = [(text_io_base, None)]);
 
+    let buffered_reader_class = vm.try_class("_io", "BufferedReader")?;
     let raw = vm.get_attribute(text_io_base.clone(), "buffer").unwrap();
 
+    if !objtype::isinstance(&raw, &buffered_reader_class) {
+        // TODO: this should be io.UnsupportedOperation error which derives both from ValueError *and* OSError
+        return Err(vm.new_value_error("not readable".to_string()));
+    }
+
     if let Ok(bytes) = vm.call_method(&raw, "read", PyFuncArgs::default()) {
         let value = objbytes::get_value(&bytes).to_vec();
 
         //format bytes into string
-        let rust_string = String::from_utf8(value).unwrap();
+        let rust_string = String::from_utf8(value).map_err(|e| {
+            vm.new_unicode_decode_error(format!(
+                "cannot decode byte at index: {}",
+                e.utf8_error().valid_up_to()
+            ))
+        })?;
         Ok(vm.ctx.new_str(rust_string))
     } else {
         Err(vm.new_value_error("Error unpacking Bytes".to_string()))
     }
 }
 
+fn text_io_base_write(vm: &VirtualMachine, args: PyFuncArgs) -> PyResult {
+    arg_check!(
+        vm,
+        args,
+        required = [(text_io_base, None), (obj, Some(vm.ctx.str_type()))]
+    );
+
+    let buffered_writer_class = vm.try_class("_io", "BufferedWriter")?;
+    let raw = vm.get_attribute(text_io_base.clone(), "buffer").unwrap();
+
+    if !objtype::isinstance(&raw, &buffered_writer_class) {
+        // TODO: this should be io.UnsupportedOperation error which derives from ValueError and OSError
+        return Err(vm.new_value_error("not writable".to_string()));
+    }
+
+    let write = vm
+        .get_method(raw.clone(), "write")
+        .ok_or_else(|| vm.new_attribute_error("BufferedWriter has no write method".to_owned()))
+        .and_then(|it| it)?;
+    let bytes = objstr::get_value(obj).into_bytes();
+
+    let len = vm.invoke(
+        write,
+        PyFuncArgs::new(vec![vm.ctx.new_bytes(bytes.clone())], vec![]),
```

Comment:
You could use `vm.invoke(write, vec![vm.ctx.new_bytes(bytes.clone())])` here, since `Into<PyArgs>` is implemented for `Vec`.